main.go

```diff
@@ -0,0 +1,706 @@
+	"github.com/edgexfoundry/go-mod-bootstrap/bootstrap/flags"
+	"github.com/edgexfoundry/go-mod-configuration/configuration"
+	"github.com/edgexfoundry/go-mod-configuration/pkg/types"
+	"github.com/edgexfoundry/go-mod-core-contracts/clients/logger"
+	"github.com/edgexfoundry/go-mod-core-contracts/models"
+	"github.com/gorilla/mux"
+	"github.com/pkg/errors"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"os/signal"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"syscall"
+	"time"
+)
+
+const (
+	serviceKey      = "rfid-llrp-inventory"
+	eventDeviceName = "rfid-llrp-inventory"
+
+	BaseConsulPath = "edgex/appservices/1.0/"
+
+	ResourceROAccessReport     = "ROAccessReport"
+	ResourceReaderNotification = "ReaderEventNotification"
+	ResourceInventoryEvent     = "InventoryEvent"
+
+	maxBodyBytes        = 100 * 1024
+	coreDataPostTimeout = 3 * time.Minute
+	eventChSz           = 100
+
+	cacheFolder  = "cache"
+	tagCacheFile = "tags.json"
+	folderPerm   = 0755 // folders require the execute flag in order to create new files
+	filePerm     = 0644
+)
+
+type inventoryApp struct {
+	edgexSdk     *appsdk.AppFunctionsSDK
+	sdkCtx       atomic.Value // *appcontext.Context
+	lgr          logWrap
+	devMu        sync.RWMutex
+	devService   llrp.DSClient
+	defaultGrp   *llrp.ReaderGroup
+	snapshotReqs chan snapshotDest
+	reports      chan reportData
+}
+
+type reportData struct {
+	report *llrp.ROAccessReport
+	info   inventory.ReportInfo
+}
+
+type snapshotDest struct {
+	w      io.Writer
+	result chan error
+}
+
+type logWrap struct {
+	logger.LoggingClient
+}
+
+type lg struct {
+	key string
+	val interface{}
+}
+
+func (lgr logWrap) errIf(cond bool, msg string, params ...lg) bool {
+	if !cond {
+		return false
+	}
+
+	if len(params) > 0 {
+		parts := make([]interface{}, len(params)*2)
+		for i := range params {
+			parts[i*2] = params[i].key
+			parts[i*2+1] = params[i].val
+		}
+		lgr.Error(msg, parts...)
+	} else {
+		lgr.Error(msg)
+	}
+
+	return true
+}
+
+func (lgr logWrap) exitIf(cond bool, msg string, params ...lg) {
+	if lgr.errIf(cond, msg, params...) {
+		os.Exit(1)
+	}
+}
+
+func (lgr logWrap) exitIfErr(err error, msg string, params ...lg) {
+	lgr.exitIf(err != nil, msg, append(params, lg{"error", err})...)
+}
+
+func main() {
+	edgexSdk := &appsdk.AppFunctionsSDK{ServiceKey: serviceKey}
+	if err := edgexSdk.Initialize(); err != nil {
+		panic(fmt.Sprintf("SDK initialization failed: %v\n", err))
+	}
+
+	lgr := logWrap{edgexSdk.LoggingClient}
+	lgr.Info("Starting.")
+
+	appSettings := edgexSdk.ApplicationSettings()
+	lgr.exitIf(appSettings == nil, "Missing application settings.")
+	cc, err := getConfigClient()
+	lgr.exitIfErr(err, "Failed to create config client.")
+
+	config, err := inventory.ParseConsulConfig(edgexSdk.LoggingClient, edgexSdk.ApplicationSettings())
+	lgr.exitIf(err != nil && !errors.Is(err, inventory.ErrUnexpectedConfigItems), fmt.Sprintf("Config parse error: %v.", err))
```

Comment:
```go
// ErrUnexpectedConfigItems is returned when the input configuration map has extra keys
// and values that are left over after parsing is complete
```

We are ignoring it because it is a warning and not an error. It signifies that the configuration may not be doing what the user wanted (eg. typo), or configuration items have been left over from an older build.

Reply:
Is it logged as a warning somewhere? If so why return as an err that is then just ignored.

Reply:
I checked and it is not logged anywhere. Thoughts on this:

```go
	config, err := inventory.ParseConsulConfig(edgexSdk.LoggingClient, edgexSdk.ApplicationSettings())
	if errors.Is(err, inventory.ErrUnexpectedConfigItems) {
		// warn on unexpected config items, but do not exit
		lgr.Warn(err.Error())
		err = nil
	}
	lgr.exitIf(err != nil, fmt.Sprintf("Config parse error: %v.", err))
```